### PR TITLE
Refactor/fe

### PR DESCRIPTION
**Pull request by leissa**
* This PR factors out various parts of thorin to use [FE](https://leissa.github.io/fe/) instead. 
* It also includes #250.
* `modules` -> `external` to obey [pitchfork](https://api.csswg.org/bikeshed/?force=1&url=https://raw.githubusercontent.com/vector-of-bool/pitchfork/develop/data/spec.bs) directory conventions

### CODE REVIEW

**Comment by leissa** [comment]
@fodinabor can you have a look at the failing macos test cases? I upgraded to `macos-latest` I your `FileCheck` magic needs an update.

**Comment by leissa** [comment]
@fodinabor , @NeuralCoder3 : anyone has the mercy to give their 2 cents on this? :)

File: thorin/fe/lexer.cpp [hunk 1]
```diff
@@ -39,16 +39,15 @@ Tok Lexer::lex() {
             return *cache;
         }
 
-        loc_.begin = ahead().pos;
-        str_.clear();
+        begin();
```
Comment (fodinabor) [inline thorin/fe/lexer.cpp:42]: `begin` might not be the best name, as it's quite heavily associated with iterators in C++. maybe prefer `start`?

File: thorin/fe/lexer.cpp [hunk 2]
```diff
@@ -39,16 +39,15 @@ Tok Lexer::lex() {
...
 
 #if defined(_WIN32) && !defined(NDEBUG) // isspace asserts otherwise
         if (accept_if([](int c) { return (c & ~0xFF) == 0 ? isspace(c) : false; })) continue;
 #else
         if (accept_if(isspace)) continue;
 #endif
-        if (accept(utf8::Err)) error(loc_, "invalid UTF-8 character");
-        if (accept(utf8::EoF)) return tok(Tag::M_eof);
+        if (accept(0)) error(loc_, "invalid UTF-8 character");
```
Comment (fodinabor) [inline thorin/fe/lexer.cpp:49]: why introduce a new magic constant when you had an enum for it?
if you mean `\0`, maybe use that?

File: thorin/util/dl.h [hunk 3]
```diff
@@ -7,8 +7,8 @@ namespace thorin::dl {
 
 std::string_view extension(); ///< `".dll"` or `".so"`
 
-void* open(const std::string& filename);
-void* get(void* handle, const std::string& symbol_name);
+void* open(const char* filename);
```
Comment (fodinabor) [inline thorin/util/dl.h:10]: can we agree on `string_view`?

**Review by fodinabor — COMMENTED** [review]
From what I can see, it looks very NFC, which is probably what you want, so, LGTM.
Just a few nits.

File: thorin/util/dl.h (hunk 3 above)
Comment (leissa) [reply thorin/util/dl.h:10]: The problem is that the APIs underneath that expect a `const char*`. So, changing to `std::string_view` means, that we would have to *convert* the original string - which **is** already a `const char*` to `std::string_view` just to *convert* this thing back to `const char*` - which is btw a bit annoying to do:
```c++
void* open(std::string_view sv) {
    std::string str(sv);
    auto file = str.c_str();
    // ...
}
```
This all seemed a bit idiotic to me. Hence, I just used `const char*`. But I do see the consistency argument. So, if you insist, I'll change it to `std::string_view` :)

Comment (fodinabor) [reply thorin/util/dl.h:10]: Aah.. Right. That's been the issue, why we're not using string view here ^^ its OK :)

**Pull request merged**